### PR TITLE
fix: do not override original vmi during eventually cycle

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
The `watchVMIForPhase` function tries to fetch the vmi waiting for a specific phase.

During the `Eventually` the original vmi variable is continuously overwritten by the retrieved value one.

This could lead to a failure of the whole `Eventually` in case of one of the `Get` before reaching the desired phase fails. This is because the later attempts will not provide to the `Get` operation the correct vmi name, Actually the later attempts will provide an empty name.
In fact, a `Get` always returns a vmi (with empty fields in case of error).

Let's store the retrieved vmi in a separate variable.

Failure example: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11685/pull-kubevirt-e2e-k8s-1.28-sig-storage/1781258067900370945

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

/kind flake

### CODE REVIEW

File: tests/libwait/wait.go [hunk 1]
```diff
@@ -156,20 +156,21 @@ func (w *Waiting) watchVMIForPhase(vmi *v1.VirtualMachineInstance) *v1.VirtualMa
 	}()
 
 	timeoutMsg := fmt.Sprintf("Timed out waiting for VMI %s to enter %s phase(s)", vmi.Name, w.phases)
+	retrievedVMI := vmi
 	// FIXME the event order is wrong. First the document should be updated
 	gomega.EventuallyWithOffset(1, func(g gomega.Gomega) v1.VirtualMachineInstancePhase {
-		vmi, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
+		retrievedVMI, err = virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
```
Comment (jean-edouard) [inline tests/libwait/wait.go:162]:
```suggestion
		retrievedVMI, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
```

File: tests/libwait/wait.go [hunk 2]
```diff
@@ -156,20 +156,21 @@ func (w *Waiting) watchVMIForPhase(vmi *v1.VirtualMachineInstance) *v1.VirtualMa
 	}()
 
 	timeoutMsg := fmt.Sprintf("Timed out waiting for VMI %s to enter %s phase(s)", vmi.Name, w.phases)
+	retrievedVMI := vmi
```
Comment (jean-edouard) [inline tests/libwait/wait.go:159]:
```suggestion

```

**Review by jean-edouard — COMMENTED** [review]
/approve

**Review by xpivarc — APPROVED** [review]

**Comment by EdDev** [comment]
> in case of one of the `Get` before reaching the desired phase fails.

We usually fail the test if the Get fails.
In which scenario it is fine to fail the Get and still continue processing the wait?

**Comment by fossedihelm** [comment]
> > in case of one of the `Get` before reaching the desired phase fails.
> 
> We usually fail the test if the Get fails. In which scenario it is fine to fail the Get and still continue processing the wait?

@EdDev I think that it is in the essence of the `Eventually` that an error could occur.
The scenario you are looking for is exactly the failure example I shared in the PR description: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11685/pull-kubevirt-e2e-k8s-1.28-sig-storage/1781258067900370945

Looking at the artifacts the vmi reached the vmi state (quite soon actually), but the eventually failed because the later `get` was providing the wrong parameters.

I don't know why the (maybe first?) Get fails, but it could happen for outside reasons, and I think we should also cover this case, instead of failing the entire lane and retesting the whole suite again.

**Comment by EdDev** [comment]
> @EdDev I think that it is in the essence of the `Eventually` that an error could occur.

Well, it contradicts all tests I encountered in this project.
The Eventually waits for the a state, but in all usages I saw so far, it did not tolerate errors like these.

The use is usually `Expect(err).NotTo(HaveOccured())`
Here it used `g.Expect(…)…` which is different and not consistent with most other usages.

**Comment by fossedihelm** [comment]
> > @EdDev I think that it is in the essence of the `Eventually` that an error could occur.
> 
> Well, it contradicts all tests I encountered in this project. The Eventually waits for the a state, but in all usages I saw so far, it did not tolerate errors like these.
> 
> The use is usually `Expect(err).NotTo(HaveOccured())` Here it used `g.Expect(…)…` which is different and not consistent with most other usages.

I agree! But I also think that there could be exceptions for specific cases.
But please, consider that here I am not changing any already existing logic.

**Pull request merged**